### ISSUE TITLE
Duplicate fbjs dependencies leading to "Error: Naming collision detected"

### ISSUE THREAD

**Issue by corbt** [facebook/react-native#5467]
This issue has been reported in #5180, #5357, #6164, #5390 and #5412 (so it's a pretty visible one!). Consolidating here to have a place to discuss solutions.

### The Issue

Errors are of the following form (although the exact file that collides isn't always the same):
```
Error building DependencyGraph:
Error: Naming collision detected: /Users/rreusser/project-name/node_modules/react-native/node_modules/react/lib/AutoFocusUtils.js collides with /Users/rreusser/project-name/node_modules/react/lib/AutoFocusUtils.js
```

It appears the problem is with the packager, which globally resolves `@providesModule` annotations, vs npm, which can install different versions of the same module at different points in the tree, leading to collisions. A lot of people are running into this now that `react-native` depends on `react`, which both depend on `fbjs`. It has also been reported with other Facebook dependencies including `flux` and `fbemitter`.

### Current Workarounds
1. Delete `node_modules`, upgrade to npm@3, and reinstall everything. npm@3 makes an effort to dedupe dependencies if the accepted versions overlap.
2. Delete `node_modules`, install `fbjs` as a top-level dependency and manually delete all nested `fbjs` instances (https://github.com/facebook/react-native/issues/5390#issuecomment-172679236)

I think option 1 is cleaner, and we should update the docs to recommend npm@3 unless we find a better short-term solution to this issue.

### Possible Solutions

1. It appears there's a file `packager/blacklist.js`, maybe this could be used to ignore the conflicting react/fbjs dependencies.
2. I don't know how much 'react-native' depends on `fbjs`, but if it isn't in many places it could be removed entirely.
3. fbjs could strip its `@providesModule` annotations in the version pushed publicly on npm and rely on relative `require` statements. Long-term I see this as being a more stable solution, because npm's nested dependencies will never play nicely with a parallel global system. I suspect we'll keep seeing issues about this pop up whenever someone includes two Facebook projects with different `fbjs` versions. But, this solution is probably impossible because of how widely used `fbjs` is with Facebook.

Other options?

**Comment by corbt** [comment]
According to https://github.com/facebook/fbjs/pull/97#issuecomment-173747938 having multiple copies of fbjs in a React Native project will be supported in 0.19. :+1:

**Comment by gre** [comment]
great news :)

**Comment by ide** [comment]
It'll probably be 0.20 since the patches haven't landed yet but yes support is coming.

**Comment by blahoink** [comment]
I've had this issue before with facebook's flux library.

Before this fix comes out, you can either update every dependency you have so they're on the same version of fbjs, or downgrade your dependencies to when they didn't use fbjs yet (for flux this was 2.0.1)

Hope this helps

**Comment by jrichardlai** [comment]
For people that are still facing this issue, a temporary can be to add `fbjs` as a dependency in your package.json, and then run a script in `postinstall` to remove the extra `fbjs` https://gist.github.com/jrichardlai/a6a36352e1b98eb2946a

**Comment by sharathprabhal** [comment]
@jrichardlai 's suggestion solved it!

**Comment by lrettig** [comment]
Ran into this just now upgrading RN 0.17.0 -> 0.18.1. In case it helps anyone else, here's the error I got:

```Error building DependencyGraph:
 Error: Naming collision detected: /Users/rettig/wip/breeze-react-native/Breeze/node_modules/fbemitter/node_modules/fbjs/lib/warning.js collides with /Users/rettig/wip/breeze-react-native/Breeze/node_modules/react-native/node_modules/fbjs/lib/warning.js
    at HasteMap._updateHasteMap (HasteMap.js:132:13)
    at HasteMap.js:103:28
    at tryCallOne (/Users/rettig/wip/breeze-react-native/Breeze/node_modules/promise/lib/core.js:37:12)
    at /Users/rettig/wip/breeze-react-native/Breeze/node_modules/promise/lib/core.js:123:15
    at flush (/Users/rettig/wip/breeze-react-native/Breeze/node_modules/asap/raw.js:50:29)
    at doNTCallback0 (node.js:428:9)
    at process._tickCallback (node.js:357:13)
```

And here's what fixed it:

1. `rm -rf node_modules/`
1. `npm i fbjs`
1. `npm i`
1. `npm i fbjs` (once more for good measure?) -- at this point I was still getting the error.
1. `find . -name 'fbjs' -print` and delete all but the top-level `fbjs`, in my case just one, `./node_modules/flux/node_modules/fbjs`
1. `rm -rf $TMPDIR/react-*`
1. `watchman watch-del-all`
1. `npm cache clean`

I'm guessing some of these steps may be extraneous and order may not matter, but in any case this got me going again.

Thanks.

**Comment by 3LP** [comment]
jrichardlai YOU DAHH MANN

**Comment by lrettig** [comment]
CC @zpao 

It gets worse. Upgrading to fbjs 0.7.0 (which was just released) causes react-native to fail every time I try to build anything. I guess fbjs 0.7.0 did something to `keyMirror`, which `react-native` depends on indirectly via `fbjs` (and there is no direct dependency). Downgrading fbjs back to 0.6.1 fixed this for me.

```
bundle: Created ReactPackager
uncaught error Error: UnableToResolveError: Unable to resolve module keyMirror from /Users/foo/foo/node_modules/react-native/Libraries/Utilities/MessageQueue.js: Invalid directory /Users/node_modules/keyMirror
    at ResolutionRequest.js:356:15
    at tryCallOne (/Users/foo/foo/node_modules/promise/lib/core.js:37:12)
    at /Users/foo/foo/node_modules/promise/lib/core.js:123:15
    at flush (/Users/foo/foo/node_modules/asap/raw.js:50:29)
    at doNTCallback0 (node.js:428:9)
    at process._tickCallback (node.js:357:13)
See logs /var/folders/rs/61p0bjp90ts_r161p5pystyh0000gn/T/react-packager.log
    at SocketClient._handleMessage (SocketClient.js:139:23)
    at BunserBuf.<anonymous> (SocketClient.js:53:42)
    at emitOne (events.js:77:13)
    at BunserBuf.emit (events.js:169:7)
    at BunserBuf.process (/Users/foo/foo/node_modules/bser/index.js:289:10)
    at /Users/foo/foo/node_modules/bser/index.js:244:12
    at doNTCallback0 (node.js:428:9)
    at process._tickCallback (node.js:357:13)
Command /bin/sh failed with exit code 1
```

**Comment by joewood** [comment]
Note - fbjs@0.6.0 may be needed as latest fbjs doesn't export the necessary modules. As per https://github.com/facebook/fbjs/pull/95. I guess this will soon all be sorted out and these hacks won't be required.

Also, rather than removing files I used a project **rn-cli.config.js** that adds the **fbjs** paths to the blacklist.

**Comment by sonu27** [comment]
@lrettig I used `npm i fbjs@0.6.1` on react 0.19
I'm using the latest node and npm (version 3.6)

**Comment by sjmueller** [comment]
@ide do you know if this landed in 0.20-rc? I didn't see a mention in the release notes.

**Comment by sjmueller** [comment]
Ok I tried to upgrade, but unfortunately I'm saddened to report that `Error: Naming collision detected` is still occurring with 0.20.0-rc1

**Comment by ide** [comment]
The fix (https://github.com/facebook/react-native/pull/5084) is still pending (some internal FB issues).

**Comment by iTonyYo** [comment]
Check [**here**](https://gist.github.com/iTonyYo/ff9d76fbb3a3e366a876) to solve this issue. :clap::+1::smiley:

**Comment by maarekj** [comment]
@iTonyYo This not work at me.

**Comment by cbrevik** [comment]
I have tried again with 0.21-rc, still not working. It's been breaking every release since 0.18 for me.

Right now I have a naming collision between `node_modules/fbjs/flow/include/CSSCore.js` and `node_modules/fbjs/lib/CSSCore.js`. None of the proposed workarounds has been successful on my project.

**Comment by chrisbianca** [comment]
I battled with this for a long time, and finally got it working by doing the folowing:

1) Adding fbjs (version 0.6.1) as a separate dependency in my package.json
2) Adding an exclude to my rn-cli.config.js file, so that the file looks roughly as follows:

```
'use strict';

const blacklist = require('react-native/packager/blacklist');

var blacklist = [
  /{root folder name}\/node_modules\/.+\/node_modules\/fbjs\/.*/
]

module.exports = {
  getBlacklistRE() {
    return blacklist('', blacklist);
  },
};
```
For those that don't already have an rn-cli.config.js file, this lives in the root directory, alongside the index.ios.js and index.android.js files.

**Comment by maarekj** [comment]
I resolved it by updating fbemitter 0.9.1 to 0.9.2

**Comment by bski006** [comment]
npm ddp fixed the issue for me

**Comment by ide** [comment]
This should be fixed in 0.22-rc.
